tests/integration/plugins/test_rust.py

```diff
@@ -127,80 +74,18 @@ def test_rust_plugin_features(new_dir):
     assert output == "hello world\n"
 
 
-def test_rust_plugin_workspace(new_dir):
+def test_rust_plugin_workspace(new_dir, datadir):
     parts_yaml = textwrap.dedent(
         """\
         parts:
           foo:
             plugin: rust
-            source: .
+            source: test_rust/workspace
```

Comment:
I like for this to be simpler than before... but what's the relation between `datadir` and this directory?

Does `datadir` change the current working directory? in the documentation the examples do use `datadir` as the path base, why not here?

Reply:
(answering both comments on this one)
`datadir` doesn't change the cwd, it creates a dir inside the test's temp dir. 
The tests use a `new_dir` fixture that changes the cwd to `tmpdir`:

```python
@pytest.fixture
def new_dir(tmpdir):
    """Change to a new temporary directory."""

    cwd = os.getcwd()
    os.chdir(tmpdir)

    yield tmpdir

    os.chdir(cwd)
```

... so the tests that do `Path("bla").write_text(...)` are actually writing into the temp dir.
I'm not 100% sure why, but I think the code in `craft_parts` is set up to work off of the cwd (possibly to mimic the behavior of the command line, like running `snapcraft` on a directory that has a `snapcraft.yaml` file.

Reply:
For comparison, here's how the test might look like with only `datadir`:

```python
def test_rust_plugin(datadir):
    parts_yaml = textwrap.dedent(
        f"""\
        parts:
          foo:
            plugin: rust
            source: {datadir}/simple
        """
    )
    parts = yaml.safe_load(parts_yaml)

    lifecycle = LifecycleManager(
        parts, application_name="test_rust_plugin", cache_dir=datadir, work_dir=datadir
    )
    actions = lifecycle.plan(Step.PRIME)

    with lifecycle.action_executor() as ctx:
        ctx.execute(actions)

    binary = Path(lifecycle.project_info.prime_dir, "bin", "rust-hello")

    output = subprocess.check_output([str(binary)], text=True)
    assert output == "hello world\n"
```

there are more explicit uses of `datadir` in the yaml and in the `LifecycleManager` constructor. While I personally prefer this approach I think it's mostly subjective, and it would be considerably different than what's currently used (it wasn't my intention to change the tests *too* much).

Reply:
why don't we flatten the tree? can the data_dir fixture operate at the method level instead of the module level?

Reply:
> can the data_dir fixture operate at the method level instead of the module level?

No, it works at the module level. So if the module has many tests that want to write the same file (like `Cargo.toml` in this case) you end up having to nest directories inside the datadir

Reply:
the amount of data written here is not much, but it could lead to a burden if the dataset to copy gets large for one test.